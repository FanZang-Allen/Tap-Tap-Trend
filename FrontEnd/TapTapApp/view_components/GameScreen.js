/* eslint-disable react-hooks/rules-of-hooks */
/* eslint-disable react-native/no-inline-styles */
import React from 'react';
import {useState, useEffect} from 'react';
import {
  View,
  Text,
  StyleSheet,
  ImageBackground,
  Image,
  ScrollView,
  Button,
} from 'react-native';
import Game from '../model/Game';
import ReviewFlatList from './ReviewFlatList';
import PostFlatList from './PostFlatList';
import {InFavourite, user_data, baseUrl, headers} from '../model/Config';

const GameScreen = ({navigation, route, gotData}) => {
  const [selectedTab, setSelectedTab] = useState(1);
  const [gameData, setgameData] = useState(null);
  const [like, setLike] = useState(InFavourite(route.params.game_id));

  function renderHeader(data) {
    return (
      <View>
        <View style={styles.header_container}>
          <Image
            source={{
              uri: data.logo,
            }}
            style={styles.game_image}
            resizeMode="contain"
          />
          <View style={styles.header}>
            <Text style={styles.game_name}>{data.name}</Text>
            <Text style={styles.provider_name}>{data.provider}</Text>
            <Text style={styles.dev_name}>Dev Onboard</Text>
          </View>
        </View>
        <View style={styles.tab_bar}>
          <View
            style={
              selectedTab === 1 ? styles.focused_tab : styles.unfocused_tab
            }>
            <Text style={styles.tab_text} onPress={() => setSelectedTab(1)}>
              Detail
            </Text>
          </View>
          <View
            style={
              selectedTab === 2 ? styles.focused_tab : styles.unfocused_tab
            }>
            <Text style={styles.tab_text} onPress={() => setSelectedTab(2)}>
              Review
            </Text>
          </View>
          <View
            style={
              selectedTab === 3 ? styles.focused_tab : styles.unfocused_tab
            }>
            <Text style={styles.tab_text} onPress={() => setSelectedTab(3)}>
              Group
            </Text>
          </View>
        </View>
      </View>
    );
  }
  function renderDetail(data) {
    let header = renderHeader(data);
    return (
      <View style={styles.container}>
        {header}
        <View style={styles.detail_container}>
          <View style={styles.info_container}>
            <Text
              style={
                styles.download_text
              }>{`Downloads: ${data.downloads}`}</Text>
            <Text
              style={
                styles.download_text
              }>{`Followers: ${data.followers}`}</Text>
            <Text style={styles.download_text}>{data.genre}</Text>
            <Text style={styles.about_header}>About</Text>
          </View>
          <View style={styles.rating_container}>
            <Image
              source={require('../assets/Favourite.png')}
              style={styles.rating_image}
              resizeMode="contain"
            />
            <Text style={styles.rating_text}>{data.rating}</Text>
          </View>
        </View>
        <View style={{flex: 1, backgroundColor: '#ffffff'}}>
          <ScrollView style={styles.about_scrollView}>
            <Text style={styles.about_text}>{data.intro}</Text>
          </ScrollView>
        </View>
        <View style={{backgroundColor: '#ffffff'}}>
          <View style={{margin: 10}}>
            <Button
              title={like ? 'Remove From Favourite' : 'Add to Favourite'}
              onPress={changeFavourite}
            />
          </View>
        </View>
      </View>
    );
  }

  if (gotData !== undefined) {
    return renderDetail(gotData);
  }

  useEffect(() => {
    let fetchCalled = false;
    async function intialFetch() {
      const gameClass = new Game(route.params.game_id);
      await gameClass.FetchData();
      if (gameClass.success === false) {
        setgameData(undefined);
      } else {
        setgameData(gameClass.GetData());
      }
    }
    if (fetchCalled === false) {
      intialFetch();
      fetchCalled = true;
    }
  }, [route.params.game_id]);

  async function changeFavourite() {
    const fetchUrl =
      baseUrl +
      `/api/user/favourite?user_id=${user_data.id}&game_id=${route.params.game_id}`;
    if (like === true) {
      try {
        let response = await fetch(fetchUrl, {
          method: 'DELETE',
          headers: headers,
        });
        let response_data = await response.json();
        if (response.status !== 200) {
          console.log(response_data.Error_message);
        } else {
          const index = user_data.favourite_list.indexOf(route.params.game_id);
          user_data.favourite_list.splice(index, 1);
          setLike(false);
          console.log(response_data.Response);
        }
      } catch (e) {
        console.log(e);
      }
    } else {
      try {
        let response = await fetch(fetchUrl, {
          method: 'PUT',
          headers: headers,
        });
        let response_data = await response.json();
        if (response.status !== 200) {
          console.log(response_data.Error_message);
        } else {
          user_data.favourite_list.push(route.params.game_id);
          setLike(true);
          console.log(response_data.Response);
        }
      } catch (e) {
        console.log(e);
      }
    }
  }

  function renderReview() {
    let header = renderHeader(gameData);
    return (
      <View style={styles.container}>
        {header}
        <ReviewFlatList
          gotData={gameData.reviews}
          navigation={navigation}
          refreshing={false}
          onRefresh={() => console.log('This screen not suppport refresh')}
        />
      </View>
    );
  }
  function renderGroup() {
    let header = renderHeader(gameData);
    return (
      <View style={styles.container}>
        {header}
        <PostFlatList
          gotData={gameData.posts}
          navigation={navigation}
          refreshing={false}
          onRefresh={() => console.log('This screen not suppport refresh')}
        />
      </View>
    );
  }
  function RenderFailed() {
    return (
      <View style={styles.container}>
        <ImageBackground
          source={{
            uri: 'https://scottleejenkins.files.wordpress.com/2014/07/werwerwerwerwerwer.jpg',
          }}
          style={styles.background}
        />
      </View>
    );
  }
  function RenderLoading() {
    return (
      <View style={styles.background}>
        <Text>Loading</Text>
      </View>
    );
  }
  if (gameData === null) {
    return RenderLoading();
  } else if (gameData === undefined) {
    return RenderFailed();
  } else {
    if (selectedTab === 1) {
      return renderDetail(gameData);
    } else if (selectedTab === 2) {
      return renderReview();
    } else {
      return renderGroup();
    }
  }
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  header_container: {
    height: 120,
    backgroundColor: '#426A70',
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'flex-start',
  },
  game_image: {
    flex: 1,
    width: 100,
    height: 100,
    margin: 10,
  },
  rating_image: {
    flex: 1,
    width: 40,
    height: 40,
    marginTop: 15,
  },
  header: {
    flex: 3,
    flexDirection: 'column',
    justifyContent: 'space-between',
    marginVertical: 10,
  },
  game_name: {
    fontSize: 23,
    fontWeight: 'bold',
    color: '#ffffff',
  },
  provider_name: {
    marginTop: 2,
    fontSize: 19,
    color: '#ffffff',
  },
  dev_name: {
    fontSize: 18,
    color: '#0B0730',
  },
  detail_container: {
    height: 160,
    backgroundColor: '#ffffff',
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'flex-start',
  },
  info_container: {
    flex: 7,
    flexDirection: 'column',
    justifyContent: 'space-between',
    marginVertical: 10,
    marginHorizontal: 10,
  },
  rating_container: {
    flex: 3,
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
  },
  download_text: {
    color: '#9B9B9B',
    fontSize: 18,
    fontWeight: 'bold',
  },
  rating_text: {
    flex: 2,
    color: '#4E40F8',
    fontSize: 30,
    fontWeight: 'bold',
  },
  background: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    paddingTop: 20,
    paddingBottom: 10,
  },
  tab_bar: {
    flexDirection: 'row',
    justifyContent: 'space-evenly',
    height: 50,
  },
  tab_text: {
    fontSize: 20,
    color: 'black',
    fontWeight: 'bold',
    alignSelf: 'center',
  },
  about_header: {
    marginTop: 5,
    color: 'black',
    fontSize: 22,
    fontWeight: 'bold',
    textDecorationLine: 'underline',
  },
  about_scrollView: {
    marginHorizontal: 10,
  },
  about_text: {
    color: 'black',
    fontSize: 16,
  },
  focused_tab: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'white',
    borderColor: 'blue',
    borderBottomWidth: 2,
  },
  unfocused_tab: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'white',
    borderColor: 'blue',
    borderBottomWidth: 0,
  },
});

export default GameScreen;
