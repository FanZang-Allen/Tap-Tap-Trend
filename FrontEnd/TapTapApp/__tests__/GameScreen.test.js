import 'react-native';
import React from 'react';
import GameScreen from '../view_components/GameScreen';
import renderer from 'react-test-renderer';

let testData = {
  downloads: '7047196',
  followers: '1295060',
  genre: 'Genre: PVP, PUBG, Online, RPG, Multiplayer, Shooter',
  id: '82354',
  intro: 'Drop.',
  logo: 'https://img.tapimg.com/market/icons/4cf4ff9eb0d623a2ae9cf5ecc49879c3_360.png?imageMogr2/auto-orient/strip',
  name: 'Fortnite',
  posts: undefined,
  posts_link: [],
  provider: 'Epic Games',
  rating: '6.9',
  region: ' Global ',
};

test('Game Rendered Page', () => {
  const snap = renderer
    .create(<GameScreen gotData={testData} route={{params: {game_id: '1'}}} />)
    .toJSON();
  expect(snap).toMatchSnapshot();
});
