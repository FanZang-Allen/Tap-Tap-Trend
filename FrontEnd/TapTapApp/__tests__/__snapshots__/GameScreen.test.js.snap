// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game Rendered Page 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "backgroundColor": "#426A70",
          "flexDirection": "row",
          "height": 120,
          "justifyContent": "space-between",
        }
      }
    >
      <Image
        resizeMode="contain"
        source={
          Object {
            "uri": "https://img.tapimg.com/market/icons/4cf4ff9eb0d623a2ae9cf5ecc49879c3_360.png?imageMogr2/auto-orient/strip",
          }
        }
        style={
          Object {
            "flex": 1,
            "height": 100,
            "margin": 10,
            "width": 100,
          }
        }
      />
      <View
        style={
          Object {
            "flex": 3,
            "flexDirection": "column",
            "justifyContent": "space-between",
            "marginVertical": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#ffffff",
              "fontSize": 23,
              "fontWeight": "bold",
            }
          }
        >
          Fortnite
        </Text>
        <Text
          style={
            Object {
              "color": "#ffffff",
              "fontSize": 19,
              "marginTop": 2,
            }
          }
        >
          Epic Games
        </Text>
        <Text
          style={
            Object {
              "color": "#0B0730",
              "fontSize": 18,
            }
          }
        >
          Dev Onboard
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "space-evenly",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderBottomWidth": 2,
            "borderColor": "blue",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          onPress={[Function]}
          style={
            Object {
              "alignSelf": "center",
              "color": "black",
              "fontSize": 20,
              "fontWeight": "bold",
            }
          }
        >
          Detail
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderBottomWidth": 0,
            "borderColor": "blue",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          onPress={[Function]}
          style={
            Object {
              "alignSelf": "center",
              "color": "black",
              "fontSize": 20,
              "fontWeight": "bold",
            }
          }
        >
          Review
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderBottomWidth": 0,
            "borderColor": "blue",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          onPress={[Function]}
          style={
            Object {
              "alignSelf": "center",
              "color": "black",
              "fontSize": 20,
              "fontWeight": "bold",
            }
          }
        >
          Group
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "backgroundColor": "#ffffff",
        "flexDirection": "row",
        "height": 160,
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 7,
          "flexDirection": "column",
          "justifyContent": "space-between",
          "marginHorizontal": 10,
          "marginVertical": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#9B9B9B",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Downloads: 7047196
      </Text>
      <Text
        style={
          Object {
            "color": "#9B9B9B",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Followers: 1295060
      </Text>
      <Text
        style={
          Object {
            "color": "#9B9B9B",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Genre: PVP, PUBG, Online, RPG, Multiplayer, Shooter
      </Text>
      <Text
        style={
          Object {
            "color": "black",
            "fontSize": 22,
            "fontWeight": "bold",
            "marginTop": 5,
            "textDecorationLine": "underline",
          }
        }
      >
        About
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 3,
          "flexDirection": "column",
          "justifyContent": "center",
        }
      }
    >
      <Image
        resizeMode="contain"
        source={
          Object {
            "process": [Function],
          }
        }
        style={
          Object {
            "flex": 1,
            "height": 40,
            "marginTop": 15,
            "width": 40,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#4E40F8",
            "flex": 2,
            "fontSize": 30,
            "fontWeight": "bold",
          }
        }
      >
        6.9
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#ffffff",
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      style={
        Object {
          "marginHorizontal": 10,
        }
      }
    >
      <View>
        <Text
          style={
            Object {
              "color": "black",
              "fontSize": 16,
            }
          }
        >
          Drop.
        </Text>
      </View>
    </RCTScrollView>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#ffffff",
      }
    }
  >
    <View
      style={
        Object {
          "margin": 10,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Add to Favourite
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
